File: src/SilksongLauncher.Launcher/app/src/main/kotlin/dev/silksong/launcher/skins/core/SkinResourceLifetimeModels.kt
package dev.silksong.launcher.skins.core

import java.util.Collections

/** Opaque authoritative runtime identity values. Stamp is correlation only, NOT SkinStamp arithmetic
 * authority. Construction/binding belongs to the future adapter; prior and target may use different bindings. */
data class ResourceLifetimePlan(val planId: String, val binding: String, val generationId: String, val stamp: String)
data class ResourceLifetimeClosure(val operationId: Long, val transactionId: String,
    val prior: ResourceLifetimePlan?, val target: ResourceLifetimePlan)
data class ResourceDisposalOperation(val operationId: Long, val closure: ResourceLifetimeClosure, val allocationId: Long)

/** Retirement checkpoint only. No cancellation, decoder-failure or teardown authority. */
class ResourceLifetimeState(
    val ledger: ResourceState = ResourceState(), val active: ResourceLifetimePlan? = null,
    val candidate: ResourceLifetimeClosure? = null, val candidateSealed: Boolean = false,
    val awaitingClones: ResourceLifetimeClosure? = null, val retirement: ResourceLifetimeClosure? = null,
    pendingDisposals: Collection<ResourceDisposalOperation> = emptyList(), val nextOperationId: Long = 1,
    val awaitingDurableCompletion: Boolean = false,
) {
    val pendingDisposals: List<ResourceDisposalOperation> = Collections.unmodifiableList(pendingDisposals.toList())
}
sealed interface ResourceLifetimeEvent {
    data class BeginCandidate(val transactionId: String, val target: ResourceLifetimePlan) : ResourceLifetimeEvent
    data class Admission(val event: ResourceEvent) : ResourceLifetimeEvent
    data class SealCandidate(val closure: ResourceLifetimeClosure) : ResourceLifetimeEvent
    /** Consume the value gate ONCE, BEFORE the caller starts visual execution. No writes are performed.
     * Failed/indeterminate outcomes stay blocked; their cleanup belongs to the next slice. */
    data class AwaitDurableCompletion(val closure: ResourceLifetimeClosure) : ResourceLifetimeEvent
    /** Trusted live verification of durable successful completion of this exact stored closure,
     * not a receipt/history lookup, command intent, or permission to commit a transaction. */
    data class DurableCompletionVerified(val closure: ResourceLifetimeClosure) : ResourceLifetimeEvent
    /** Exact acknowledgement of InvalidateCompanionClones: ALL prior binding/stamp clones destroyed. */
    data class CompanionClonesInvalidated(val closure: ResourceLifetimeClosure) : ResourceLifetimeEvent
    /** Authoritative live registry count for the exact pending prior binding/stamp, not an estimate. */
    data class RegisteredCloneCountVerified(val closure: ResourceLifetimeClosure, val count: Long) : ResourceLifetimeEvent
    data class DisposeAcknowledged(val operation: ResourceDisposalOperation) : ResourceLifetimeEvent
}
sealed interface ResourceLifetimeCommand {
    data class Admission(val command: ResourceCommand) : ResourceLifetimeCommand
    data class InvalidateCompanionClones(val closure: ResourceLifetimeClosure) : ResourceLifetimeCommand
    data class Dispose(val operation: ResourceDisposalOperation) : ResourceLifetimeCommand
}
class ResourceLifetimeDecision(val state: ResourceLifetimeState, val accepted: Boolean, val diagnosis: String,
    commands: Collection<ResourceLifetimeCommand> = emptyList()) {
    val commands: List<ResourceLifetimeCommand> = Collections.unmodifiableList(commands.toList())
}

File: tools/hollow-knight-patches/src/skins/core/SkinResourceLifetimeModels.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DualSouls.Skins.HollowKnight.Core
{
    // Opaque authoritative runtime identities. Stamp is correlation only, NOT SkinStamp arithmetic authority.
    // Construction/binding belongs to the future adapter; prior and target may use different bindings.
    public sealed record ResourceLifetimePlan(string PlanId, string Binding, string GenerationId, string Stamp);
    public sealed record ResourceLifetimeClosure(long OperationId, string TransactionId, ResourceLifetimePlan Prior, ResourceLifetimePlan Target);
    public sealed record ResourceDisposalOperation(long OperationId, ResourceLifetimeClosure Closure, long AllocationId);
    // Retirement checkpoint only. No cancellation, decoder-failure or teardown authority.
    public sealed class ResourceLifetimeState
    {
        public ResourceState Ledger { get; }
        public ResourceLifetimePlan Active { get; }
        public ResourceLifetimeClosure Candidate { get; }
        public bool CandidateSealed { get; }
        public bool AwaitingDurableCompletion { get; }
        public ResourceLifetimeClosure AwaitingClones { get; }
        public ResourceLifetimeClosure Retirement { get; }
        public IReadOnlyList<ResourceDisposalOperation> PendingDisposals { get; }
        public long NextOperationId { get; }
        public ResourceLifetimeState(ResourceState ledger = null, ResourceLifetimePlan active = null,
            ResourceLifetimeClosure candidate = null, bool candidateSealed = false,
            ResourceLifetimeClosure awaitingClones = null, ResourceLifetimeClosure retirement = null,
            IEnumerable<ResourceDisposalOperation> pendingDisposals = null, long nextOperationId = 1, bool awaitingDurableCompletion = false)
        {
            Ledger = ledger ?? new ResourceState(); Active = active; Candidate = candidate; CandidateSealed = candidateSealed;
            AwaitingClones = awaitingClones; Retirement = retirement; NextOperationId = nextOperationId; AwaitingDurableCompletion = awaitingDurableCompletion;
            PendingDisposals = Array.AsReadOnly((pendingDisposals ?? Array.Empty<ResourceDisposalOperation>()).ToArray());
        }
    }
    public abstract record ResourceLifetimeEvent
    {
        private ResourceLifetimeEvent() { }
        public sealed record BeginCandidate(string TransactionId, ResourceLifetimePlan Target) : ResourceLifetimeEvent;
        public sealed record Admission(ResourceEvent Event) : ResourceLifetimeEvent;
        public sealed record SealCandidate(ResourceLifetimeClosure Closure) : ResourceLifetimeEvent;
        // Consume the value gate ONCE, BEFORE the caller starts visual execution. Performs no writes.
        // Failed/indeterminate outcomes stay blocked; cleanup belongs to the next slice.
        public sealed record AwaitDurableCompletion(ResourceLifetimeClosure Closure) : ResourceLifetimeEvent;
        // Trusted LIVE verification of durable successful completion of the exact stored closure;
        // NOT receipt/history lookup, command intent, or authority to commit a transaction.
        public sealed record DurableCompletionVerified(ResourceLifetimeClosure Closure) : ResourceLifetimeEvent;
        // Exact InvalidateCompanionClones acknowledgement: ALL prior binding/stamp clones destroyed.
        public sealed record CompanionClonesInvalidated(ResourceLifetimeClosure Closure) : ResourceLifetimeEvent;
        // Authoritative live registry count for the exact pending prior binding/stamp, not an estimate.
        public sealed record RegisteredCloneCountVerified(ResourceLifetimeClosure Closure, long Count) : ResourceLifetimeEvent;
        public sealed record DisposeAcknowledged(ResourceDisposalOperation Operation) : ResourceLifetimeEvent;
    }
    public abstract record ResourceLifetimeCommand
    {
        private ResourceLifetimeCommand() { }
        public sealed record Admission(ResourceCommand Command) : ResourceLifetimeCommand;
        public sealed record InvalidateCompanionClones(ResourceLifetimeClosure Closure) : ResourceLifetimeCommand;
        public sealed record Dispose(ResourceDisposalOperation Operation) : ResourceLifetimeCommand;
    }
    public sealed class ResourceLifetimeDecision
    {
        public ResourceLifetimeState State { get; }
        public bool Accepted { get; }
        public string Diagnosis { get; }
        public IReadOnlyList<ResourceLifetimeCommand> Commands { get; }
        public ResourceLifetimeDecision(ResourceLifetimeState state, bool accepted, string diagnosis, params ResourceLifetimeCommand[] commands)
        {
            State = state; Accepted = accepted; Diagnosis = diagnosis;
            Commands = Array.AsReadOnly((ResourceLifetimeCommand[])commands.Clone());
        }
    }
}
